csharp/src/Google.Protobuf/CodedInputStream.cs

```diff
@@ -1002,19 +1526,31 @@ internal bool ReachedLimit
         /// </summary>
         public bool IsAtEnd
         {
-            get { return bufferPos == bufferSize && !RefillBuffer(false); }
+            [SecuritySafeCritical]
+            get
+            {
+                var immediateBudder = ImmediateBuffer;
+                return IsAtEndCore(ref immediateBudder);
+            }
         }
 
+        /// <summary>
+        /// This supports the Protocol Buffers infrastructure and is not meant to be used directly from your code.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [SecurityCritical]
+        internal bool IsAtEndCore(ref ReadOnlySpan<byte> immediateBuffer) => bufferPos == bufferSize && (input == null && !hasNativeInput || !RefillBuffer(false, ref immediateBuffer));

```

Comment:
nit: why "Core"? how about "Impl"?

Reply:
In C# I have seen more Core than Impl (as in java and C++)